File: 2021/day06/day6.cpp
#include <iostream>
#include <vector>
#include <string>
#include <sstream>
#include <algorithm>
#include <numeric>

using namespace std;
typedef long long ll;

void part1(vector<ll> ishCount);
void part2(vector<ll> fishCount);
vector<ll> readInput();
vector<ll> simulate1Day(vector<ll> fishCount);
vector<ll> simulateNDays(int n, vector<ll> fishCount);
ll getNumberOfFish(vector<ll> fishCount);


int main() {
    vector<ll> fishCount = readInput();
    part1(fishCount);
    part2(fishCount);
    return 0;
}

void part1(vector<ll> fishCount) {
    vector<ll> updatedFishCount = simulateNDays(80, fishCount);
    ll numberOfFish = getNumberOfFish(updatedFishCount);
    cout << "Part 1: " << numberOfFish << endl;
}

void part2(vector<ll> fishCount) {
    vector<ll> updatedFishCount = simulateNDays(256, fishCount);
    ll numberOfFish = getNumberOfFish(updatedFishCount);
    cout << "Part 2: " << numberOfFish << endl;
}

vector<ll> readInput() {
    string rawInput; cin >> rawInput;
    vector<ll> fishCount(9, 0);
    stringstream ss(rawInput);
    for (int i; ss >> i;) {
        fishCount[i]++;
        if (ss.peek() == ',') {
            ss.ignore();
        }
    }
    return fishCount;
}

vector<ll> simulate1Day(vector<ll> fishCount) {
    ll fishDue = fishCount[0];
    for (int i = 1; i < 9; i++) {
        fishCount[i - 1] = fishCount[i];
    }
    fishCount[8] = fishDue;
    fishCount[6] += fishDue;
    return fishCount;
}

vector<ll> simulateNDays(int n, vector<ll> fishCount) {
    for (int i = 0; i < n; i++) {
        vector<ll> updatedFishCount = simulate1Day(fishCount);
        fishCount = updatedFishCount;
    }
    return fishCount;
}

ll getNumberOfFish(vector<ll> fishCount) {
    return accumulate(fishCount.begin(), fishCount.end(), 0ll);
}

File: 2021/day06/day6.py
def read_input():
    file = open('input.in', 'r')
    initial_count = list(map(int, file.readline().strip().split(",")))
    fish_count = [0]*9
    for fish in initial_count:
        fish_count[fish] += 1
    return fish_count


def simulate_1_day(fish_count):
    fish_due = fish_count[0]
    for i in range(1, 9):
        fish_count[i - 1] = fish_count[i]
    fish_count[8] = fish_due
    fish_count[6] += fish_due
    return fish_count


def simulate_n_days(n, fish_count):
    for _ in range(n):
        fish_count = simulate_1_day(fish_count)
    return fish_count


def get_number_of_fish(fish_count):
    return sum(fish_count)


def part1():
    initial_count = read_input()
    fish_count = simulate_n_days(80, initial_count)
    number_of_fish = get_number_of_fish(fish_count)
    print(f"Part 1: {number_of_fish}")


def part2():
    fish_count = read_input()
    fish_count = simulate_n_days(256, fish_count)
    number_of_fish = get_number_of_fish(fish_count)
    print(f"Part 2: {number_of_fish}")


if __name__ == "__main__":
    part1()
    part2()
